File: tutocpp_pkg/src/client_node.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/srv/add_two_ints.hpp"

class ClientNode : public rclcpp::Node
{
public:
    ClientNode() : Node("client_node")
    {
        thread1_ = std::thread(std::bind(&ClientNode::callServer, this, 6, 7));
        RCLCPP_INFO(get_logger(), "Client Node has been started");
    }

    void callServer(int a, int b)
    {
        auto client = create_client<example_interfaces::srv::AddTwoInts>("add_ints");
        while (!client->wait_for_service(std::chrono::seconds(1)))
        {
            RCLCPP_WARN(get_logger(), "Waiting for the server to be up...");
        }
        auto request = std::make_shared<example_interfaces::srv::AddTwoInts::Request>();
        request->a = a;
        request->b = b;

        auto future = client->async_send_request(request);
        try
        {
            auto response = future.get();
            RCLCPP_INFO(get_logger(), "%ld + %ld = %ld", request->a, request->b, response->sum);
        }
        catch (const std::exception &e)
        {
            RCLCPP_ERROR(get_logger(), e.what());
        }
    }
private:
    std::thread thread1_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);

    auto node = std::make_shared<ClientNode>();

    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: tutopy_pkg/tutopy_pkg/client_node.py
#!/usr/bin/env python3
import rclpy

from example_interfaces.srv import AddTwoInts
from functools import partial
from rclpy.node import Node


class ClientNode(Node):
    def __init__(self):
        super().__init__("client_node")
        self.call_server(6, 7)

    def call_server(self, a, b):
        client = self.create_client(AddTwoInts, "add_ints")
        while not client.wait_for_service(1.0):
            self.get_logger().warn("Waiting for Server")

        request = AddTwoInts.Request()
        request.a = a
        request.b = b
        future = client.call_async(request)
        future.add_done_callback(partial(self.callback_done, a=a, b=b))

    def callback_done(self, future, a, b):
        try:
            response = future.result()
            self.get_logger().info(f"{a} + {b} = {response.sum}")
        except Exception as e:
            self.get_logger().error(f"Service call failed: {e}")


def main(args=None):
    rclpy.init(args=args)

    node = ClientNode()

    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()
